---
title: "DivRatesMS"
author: "Roi Maor"
date: "16/02/2021"
output: html_document
---
### Technical note
#### Files used in the data processing:
1 - Updating taxonomy.r
2 - Ch3_MuHiSSE.r
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Methods
### Data
#### Taxonomy and phylogeny
1 - Using the Upham 2019 phylogeny requiers updating to the Burgin 2018 taxonomy
2 - Species 'lumping' in the new taxonomy caused redundancy (duplicate records for the same recognised species). These records did not agree in two species (Monodelphis americana- diurnal and nocturnal; Heterogeomys dariensis- cathemeral and nocturnal) which were consequently assigned 'cathemeral'.
3 - There is still some mismatch between MDD (all versions) and the phylogeny. 
3.1 - As a rule, I changed the name in the phylogeny to conform to MDD_v1, unless MDD_v1.31 (the most recent to date) agrees with the phylogeny, in which case the name in the phylogeny remained.
4 - Surprisingly the phylogeny had more mismatches with MDD names than with MSW3 names in my data, perhaps reflecting that the phylogenetic work was started much sooner (perhaps retaining names from old sequencing studies) than the taxonomic revision

5 - The bottom line: the column "data$Phylo_name" is theone to use for matching data to the phylogeny binomials
```{r}
summary(data$AP)
summary(data)

```

### Models

CID, diversification increased due to KPg : BiSSE.td (MuSSE.td?)





------

* METHODS considerations
1 - 
Each hidden state set (i.e. all states of A, or all states of B etc.) represents a trait that affects diversification.
For example, my focal trait is AP, so trait dependent diversification will not capture diversification increases due to flying habits (bats), gulp feeding (whales), granivory (rodents) etc., but all these cannot be reliably model by a single hidden state, and will require multiple hidden states. 

2 - 
Use BOTH the MCC tree (based on ALL posterior trees) AND a range of trees (individually) from the posterior

3 - 
There should not be more hidden rates than there are hidden states! (see my comment on Nakov 2019, p. 2500)

4 - 
I don't think there should be equal transition rates AMONG hidden states, because symmetrical backward-forward transitions make little sense in a trait that affects diversification 

5 - 
All hidden rates different (i.e. 6  transition rates between every 2 hidden states) makes no sense because it means that hidden rates vary depending on observed state (=oxymoron) (e.g. if 1A -> 1B != 2A -> 2B)

6 - 
Meseguer et al. 2020, J Biogeog:

"To account for rate variation, we compared models where diversification rates were allowed to change at 
distinct time intervals (BiSSE.td). For the latter, we implemented one time shift (two time intervals) 
and tested different shift times at 3.5, 15, 34 and 66 Ma, which represent periods of environmental
changes potentially affecting diversification: 3.5 Ma corresponds to the final closure of the Panama 
Isthmus and the initiation of Northern Hemisphere glaciations (Zachos, Dickens, & Zeebe, 2008); 15 Ma 
coincides with the first land connections between South and Central Americas (Montes et al., 2015) and 
the middle Miocene climatic optimum (Zachos et al., 2008); 34 Ma is the opening of the Drake passage (and 
collateral end of Antarctica–South America connection) with the onset of the Antarctic Circumpolar Current 
(Scher et al., 2015) and the GAARlandia bridge between North and South Americas (Iturralde-Vinent & MacPhee, 
1999); 66 Ma is the Cretaceous–Palaeogene mass extinction (Longrich, Bhullar, & Gauthier, 2012). Implementing 
models with more than one time shift resulted in computationally (20 parameters) intractable analyses for 
phylogenies with thousands of species." 

"For a character state to be considered a character in state speciation and extinction (SSE) models, it has 
to be represented in >10% of the species, because the power of the trait-dependent models has been shown to 
be severely affected when high tip ratio bias is observed (Alves, Diniz-Filho, & Villalobos, 2017; Davis, 
Midford, & Maddison, 2013)."


```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.


```{r}

```
